src/language_features/document_links.rs

```diff
@@ -0,0 +1,213 @@
+            )
+        })
+        .join(" ");
+
+    ctx.document_links.insert(buffile.clone(), links);
+
+    let command = format!("set-option buffer lsp_document_links {version} {range_specs}");
+    let command = format!(
+        "evaluate-commands -buffer {} %§{}§",
+        editor_quote(buffile),
+        command.replace('§', "§§")
+    );
+    ctx.exec(EditorMeta::default(), command);
+}
+
+#[derive(Clone, Debug)]
+pub struct DocumentLinkOptions {
+    pub selection_desc: String,
+}
+
+pub fn resolve_and_follow_document_link(
+    meta: EditorMeta,
+    params: DocumentLinkOptions,
+    ctx: &mut Context,
+) {
+    let (range, _cursor) = parse_kakoune_range(&params.selection_desc);
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => return,
+    };
+
+    if let Some((server_id, link)) = ctx
+        .document_links
+        .get(&meta.buffile)
+        .and_then(|links| {
+            links.iter().find(|(server_id, link)| {
+                let ServerSettings {
+                    offset_encoding, ..
+                } = ctx.server(*server_id);
+                let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+                ranges_overlap(link.range, range)
+            })
+        })
+        .filter(|(_, link)| link.target.is_none())
+        .cloned()
+    {
+        let mut req_params = HashMap::new();
+        req_params.insert(server_id, vec![link]);
+
+        ctx.call::<DocumentLinkResolve, _>(
+            meta,
+            RequestParams::Each(req_params),
+            |ctx: &mut Context, meta, results| follow_document_link(meta, &results, ctx),
+        );
+        return;
+    }
+
+    let no_links = vec![];
+    let links = ctx.document_links.get(&meta.buffile).unwrap_or(&no_links);
+    let mut links = links
+        .iter()
+        .filter(|(server_id, link)| {
+            let ServerSettings {
+                offset_encoding, ..
+            } = ctx.server(*server_id);
+            let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+            ranges_overlap(link.range, range)
+        })
+        .map(|(a, b)| (*a, b.clone()))
+        .collect::<Vec<_>>();
+
+    links.sort_by_key(|(_server_name, link)| {
+        let Range { start, end } = link.range;
+        end.line - start.line
+    });
+
+    follow_document_link(meta, &links, ctx);
+}
+
+fn follow_document_link(meta: EditorMeta, links: &[(ServerId, DocumentLink)], ctx: &mut Context) {
+    let commands = links
+        .iter()
+        .filter(|(_, link)| link.target.is_some())
+        .map(|(_, link)| {
+            let target = link.target.as_ref().unwrap();
+            let label = editor_quote(target.as_str());
+            let path = Some(target)
+                .filter(|x| x.scheme() == "file")
+                .and_then(|x| x.to_file_path().ok())
+                .map(|x| x.to_string_lossy().to_string());
```

Comment:
super nit:

I guess our error handling is all over the place :(
If we started a project from scratch we'd definitely add an easy way to log
these errors, since they probably indicate a bug in the LSP server.

I guess if `to_string_lossy` is forced to do lossy conversion, we at least
get part of the name, so the user will see a somewhat reasonable error message.
But logging and propagating some error to Kakoune seems like it would be ideal,
because AFAICT, Kakoune will not be able to open the right file if we put
the Unicode replacement character in there.